cmd/codegen/generator/go/templates/modules.go

```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
+			if !ok {
+				return nil, fmt.Errorf("expected struct type for %s", param.Name())
 			}
-		} else {
-			argTypeDef, err := ps.goTypeToAPIType(param.Type(), nil)
-			if err != nil {
-				return nil, fmt.Errorf("failed to convert param type: %w", err)
+
+			parent := &paramSpec{
+				name:      params.At(i).Name(),
+				paramType: param.Type(),
+				baseType:  param.Type(),
 			}
 
-			fnDef = dotLine(fnDef, "WithArg").Call(Lit(param.Name()), argTypeDef)
+			for f := 0; f < paramType.NumFields(); f++ {
+				spec, err := ps.parseParamSpecVar(paramType.Field(f))
+				if err != nil {
+					return nil, err
+				}
+				spec.parent = parent
+				spec.description = stype.Fields.List[f].Doc.Text()
+				if spec.description == "" {
+					spec.description = stype.Fields.List[f].Comment.Text()
+				}
+				spec.description = strings.TrimSpace(spec.description)
+				specs = append(specs, spec)
+			}
+			return specs, nil
 		}
 	}
 
-	return fnDef, nil
+	for ; i < params.Len(); i++ {
+		spec, err := ps.parseParamSpecVar(params.At(i))
+		if err != nil {
+			return nil, err
+		}
+		if sig.Variadic() && i == params.Len()-1 {
+			spec.variadic = true
+		}
+
+		pos := fnDecl.Type.Params.List[i].Pos()
+		if cmt, err := ps.commentForPos(pos - 1); err == nil {
+			spec.description = cmt.Text()
+			spec.description = strings.TrimSpace(spec.description)
+		}
+
+		specs = append(specs, spec)
+	}
+	return specs, nil
 }
 
-func namedOrDirectStruct(t types.Type) (*types.Struct, bool) {
-	switch x := t.(type) {
-	case *types.Named:
-		return namedOrDirectStruct(x.Underlying())
-	case *types.Struct:
-		return x, true
-	default:
-		return nil, false
+func (ps *parseState) parseParamSpecVar(field *types.Var) (paramSpec, error) {
+	if _, ok := field.Type().(*types.Struct); ok {
+		return paramSpec{}, fmt.Errorf("nested structs are not supported")
+	}
+
+	paramType := field.Type()
+	baseType := paramType
+	for {
+		ptr, ok := baseType.(*types.Pointer)
+		if !ok {
+			break
+		}
+		baseType = ptr.Elem()
+	}
+
+	optional := false
+	if named, ok := baseType.(*types.Named); ok {
+		if named.Obj().Name() == "Optional" {
+			typeArgs := named.TypeArgs()
+			if typeArgs.Len() != 1 {
+				return paramSpec{}, fmt.Errorf("optional type must have exactly one type argument")
+			}
+			optional = true
+
+			baseType = typeArgs.At(0)
+			for {
+				ptr, ok := baseType.(*types.Pointer)
+				if !ok {
+					break
+				}
+				baseType = ptr.Elem()
+			}
+		}
 	}
+
+	return paramSpec{
+		name:      field.Name(),
+		paramType: paramType,
+		baseType:  baseType,
+		optional:  optional,
+	}, nil
+}
+
+type paramSpec struct {
+	name        string
+	description string
+
+	// paramType is the full type declared in the function signature, which may
+	// include pointer types, Optional, etc
+	paramType types.Type
+	// baseType is the simplified base type derived from the function signature
+	baseType types.Type
+
+	optional bool
+	variadic bool
+
+	defaultValue string
+
+	parent *paramSpec
```

Comment:
nit: quick comment on what `parent` is would help, otherwise have to dig through code to grok the case it's used for